diff --git a/tools/releasetools/edify_generator.py b/tools/releasetools/edify_generator.py
index de82455..6f71014 100644
--- a/tools/releasetools/edify_generator.py
+++ b/tools/releasetools/edify_generator.py
@@ -98,7 +98,7 @@ class EdifyGenerator(object):
            ' || abort("This package is for \\"%s\\" devices; '
            'this is a \\"" + getprop("ro.product.device") + "\\".");'
            ');') % device
-    self.script.append(self._WordWrap(cmd))
+    #self.script.append(self._WordWrap(cmd))
 
   def AssertSomeBootloader(self, *bootloaders):
     """Asert that the bootloader version is one of *bootloaders."""
